File: packages/capacitor-notificare-user-inbox/android/src/main/java/re/notifica/inbox/user/capacitor/NotificareUserInboxPlugin.kt
package re.notifica.inbox.user.capacitor

import com.getcapacitor.*
import com.getcapacitor.annotation.CapacitorPlugin
import re.notifica.Notificare
import re.notifica.NotificareCallback
import re.notifica.inbox.user.ktx.userInbox
import re.notifica.inbox.user.models.NotificareUserInboxItem
import re.notifica.models.NotificareNotification


@CapacitorPlugin(name = "NotificareUserInboxPlugin")
public class NotificareUserInboxPlugin : Plugin() {
    @PluginMethod
    public fun parseResponseFromJson(call: PluginCall) {
        val json = call.getObject("json") ?: run {
            call.reject("Missing 'json' parameter.")
            return
        }

        try {
            val response = Notificare.userInbox().parseResponse(json)
            call.resolve(
                JSObject().apply {
                    put("result", response.toJson())
                }
            )
        } catch (e: Exception) {
            call.reject(e.localizedMessage)
        }
    }

    @PluginMethod
    public fun parseResponseFromString(call: PluginCall) {
        val json = call.getString("json") ?: run {
            call.reject("Missing 'json' parameter.")
            return
        }

        try {
            val response = Notificare.userInbox().parseResponse(json)
            call.resolve(
                JSObject().apply {
                    put("result", response.toJson())
                }
            )
        } catch (e: Exception) {
            call.reject(e.localizedMessage)
        }
    }

    @PluginMethod
    public fun open(call: PluginCall) {
        val json = call.getObject("item") ?: run {
            call.reject("Missing 'item' parameter.")
            return
        }

        val item: NotificareUserInboxItem = try {
            NotificareUserInboxItem.fromJson(json)
        } catch (e: Exception) {
            call.reject(e.localizedMessage)
            return
        }

        Notificare.userInbox().open(item, object : NotificareCallback<NotificareNotification> {
            override fun onSuccess(result: NotificareNotification) {
                try {
                    call.resolve(
                        JSObject().apply {
                            put("result", result.toJson())
                        }
                    )
                } catch (e: Exception) {
                    call.reject(e.localizedMessage)
                }
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    @PluginMethod
    public fun markAsRead(call: PluginCall) {
        val json = call.getObject("item") ?: run {
            call.reject("Missing 'item' parameter.")
            return
        }

        val item: NotificareUserInboxItem = try {
            NotificareUserInboxItem.fromJson(json)
        } catch (e: Exception) {
            call.reject(e.localizedMessage)
            return
        }

        Notificare.userInbox().markAsRead(item, object : NotificareCallback<Unit> {
            override fun onSuccess(result: Unit) {
                call.resolve()
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    @PluginMethod
    public fun remove(call: PluginCall) {
        val json = call.getObject("item") ?: run {
            call.reject("Missing 'item' parameter.")
            return
        }

        val item: NotificareUserInboxItem = try {
            NotificareUserInboxItem.fromJson(json)
        } catch (e: Exception) {
            call.reject(e.localizedMessage)
            return
        }

        Notificare.userInbox().remove(item, object : NotificareCallback<Unit> {
            override fun onSuccess(result: Unit) {
                call.resolve()
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }
}

File: packages/capacitor-notificare-user-inbox/ios/Plugin/NotificareUserInboxPlugin.swift
import Foundation
import Capacitor
import NotificareKit
import NotificareUserInboxKit

@objc(NotificareUserInboxPlugin)
public class NotificareUserInboxPlugin: CAPPlugin {
    @objc func parseResponseFromJson(_ call: CAPPluginCall) {
        guard let json = call.getObject("json") else {
            call.reject("Missing 'json' parameter.")
            return
        }
        
        do {
            let response = try Notificare.shared.userInbox().parseResponse(json: json)
            
            call.resolve([
                "result": try response.toJson()
            ])
        } catch {
            call.reject(error.localizedDescription)
        }
    }
    
    @objc func parseResponseFromString(_ call: CAPPluginCall) {
        guard let json = call.getString("json") else {
            call.reject("Missing 'json' parameter.")
            return
        }
        
        do {
            let response = try Notificare.shared.userInbox().parseResponse(string: json)
            
            call.resolve([
                "result": try response.toJson()
            ])
        } catch {
            call.reject(error.localizedDescription)
        }
    }
    
    @objc func open(_ call: CAPPluginCall) {
        guard let json = call.getObject("item") else {
            call.reject("Missing 'item' parameter.")
            return
        }
        
        let item: NotificareUserInboxItem
        
        do {
            item = try NotificareUserInboxItem.fromJson(json: json)
        } catch {
            call.reject(error.localizedDescription)
            return
        }
        
        Notificare.shared.userInbox().open(item) { result in
            switch result {
            case let .success(notification):
                do {
                    call.resolve([
                        "result": try notification.toJson()
                    ])
                } catch {
                    call.reject(error.localizedDescription)
                }
            case let .failure(error):
                call.reject(error.localizedDescription)
            }
        }
    }
    
    @objc func markAsRead(_ call: CAPPluginCall) {
        guard let json = call.getObject("item") else {
            call.reject("Missing 'item' parameter.")
            return
        }
        
        let item: NotificareUserInboxItem
        
        do {
            item = try NotificareUserInboxItem.fromJson(json: json)
        } catch {
            call.reject(error.localizedDescription)
            return
        }
        
        Notificare.shared.userInbox().markAsRead(item) { result in
            switch result {
            case .success:
                call.resolve()
            case let .failure(error):
                call.reject(error.localizedDescription)
            }
        }
    }
    
    @objc func remove(_ call: CAPPluginCall) {
        guard let json = call.getObject("item") else {
            call.reject("Missing 'item' parameter.")
            return
        }
        
        let item: NotificareUserInboxItem
        
        do {
            item = try NotificareUserInboxItem.fromJson(json: json)
        } catch {
            call.reject(error.localizedDescription)
            return
        }
        
        Notificare.shared.userInbox().remove(item) { result in
            switch result {
            case .success:
                call.resolve()
            case let .failure(error):
                call.reject(error.localizedDescription)
            }
        }
    }
}
